packages/cw721/src/testing/contract_tests.rs

```diff
@@ -0,0 +1,1547 @@
+
+    // list the token_ids
+    let tokens = contract
+        .query_all_tokens(deps.as_ref(), env, None, None)
+        .unwrap();
+    assert_eq!(1, tokens.tokens.len());
+    assert_eq!(vec![token_id], tokens.tokens);
+}
+
+#[test]
+fn mint_with_metadata() {
+    // case 1: mint with valid metadata
+    {
+        let mut deps = mock_dependencies();
+        let contract = setup_contract(deps.as_mut());
+
+        let token_id = "1".to_string();
+        let token_uri = "ipfs://foo.bar".to_string();
+        let mint_msg = Cw721ExecuteMsg::Mint {
+            token_id: token_id.clone(),
+            owner: String::from("medusa"),
+            token_uri: Some(token_uri.clone()),
+            extension: Some(Metadata {
+                image: Some("ipfs://foo.bar/image.png".to_string()),
+                image_data: Some("image data".to_string()),
+                external_url: Some("https://github.com".to_string()),
+                description: Some("description".to_string()),
+                name: Some("name".to_string()),
+                attributes: Some(vec![Trait {
+                    trait_type: "trait_type".to_string(),
+                    value: "value".to_string(),
+                    display_type: Some("display_type".to_string()),
+                }]),
+                background_color: Some("background_color".to_string()),
+                animation_url: Some("ssl://animation_url".to_string()),
+                youtube_url: Some("file://youtube_url".to_string()),
+            }),
+        };
+
+        let info_minter = mock_info(MINTER_ADDR, &[]);
+        let env = mock_env();
+        contract
+            .execute(deps.as_mut(), env.clone(), info_minter, mint_msg)
+            .unwrap();
+    }
+    // case 2: mint with invalid metadata
+    {
+        let mut deps = mock_dependencies();
+        let contract = setup_contract(deps.as_mut());
+
+        let token_id = "1".to_string();
+        let token_uri = "ipfs://foo.bar".to_string();
+        let info_minter = mock_info(MINTER_ADDR, &[]);
+        let env = mock_env();
+
+        let valid_metadata = Metadata {
+            image: Some("ipfs://foo.bar/image.png".to_string()),
+            image_data: Some("image data".to_string()),
+            external_url: Some("https://github.com".to_string()),
+            description: Some("description".to_string()),
+            name: Some("name".to_string()),
+            attributes: Some(vec![Trait {
+                trait_type: "trait_type".to_string(),
+                value: "value".to_string(),
+                display_type: Some("display_type".to_string()),
+            }]),
+            background_color: Some("background_color".to_string()),
+            animation_url: Some("ssl://animation_url".to_string()),
+            youtube_url: Some("file://youtube_url".to_string()),
+        };
+
+        // invalid image
+        let mut metadata = valid_metadata.clone();
+        metadata.image = Some("invalid".to_string());
```

Comment:
deep happy path tests down to onchain metadata level! Like it checks whether `metadata.image` has an invalid URL!